src/metric_family.h

```diff
@@ -0,0 +1,145 @@
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <string>
+#include "ipc_message.h"
+#include "metric.h"
+#include "pb_string.h"
+#include "pb_utils.h"
+
+#ifdef TRITON_PB_STUB
+#include <pybind11/embed.h>
+namespace py = pybind11;
+#else
+#include "triton/core/tritonserver.h"
+#endif
+
+namespace triton { namespace backend { namespace python {
+
+enum MetricFamilyRequestKind { MetricFamilyNew, MetricFamilyDelete };
+
+struct MetricFamilyShm {
+  bi::managed_external_buffer::handle_t name_shm_handle;
+  bi::managed_external_buffer::handle_t description_shm_handle;
+  bi::managed_external_buffer::handle_t key_name_shm_handle;
+  MetricKind kind;
+  MetricFamilyRequestKind metric_family_request_kind;
+  void* metric_family_address;
+};
+
+class MetricFamily {
+ public:
+  MetricFamily(
+      const std::string& name, const std::string& description,
+      const MetricKind& kind);
+
+  ~MetricFamily();
+
+  /// Get the name of the custom metric family.
+  /// \return Returns the name of the metric family.
+  const std::string& Name();
+
+  /// Get the description of the custom metric family.
+  /// \return Returns the description of the custom metric family.
+  const std::string& Description();
+
+  /// Get the metric kind of the custom metric family.
+  /// \return Returns the metric kind of the custom metric family.
+  const MetricKind& Kind();
+
+  /// Get the shm handle of the custom metric family.
+  /// \return Returns the shm handle of the custom metric family.
+  bi::managed_external_buffer::handle_t ShmHandle();
+
+  /// Get the custom metric family request kind.
+  /// \return Returns the custom metric family request kind.
+  const MetricFamilyRequestKind& RequestKind();
```

Comment:
This `MetricFamilyRequestKind` is used to indicate the action PB should take when communicating over shared memory, right? My understanding is that's why its part of `MetricFamilyShm`.

What is its purpose as a member of `MetricFamily` class?